idp/authmiddleware/authmiddleware.go

```diff
@@ -0,0 +1,202 @@
+// provide any information apart from the e-mail address and a the reserved group ID '3E093BE5-CCCE-435D-99F8-544656B98681'
+// which marks the user as an external user which is unknown to the system. This group ID can be used to
+// distinguish external from internal users.
+// If you are unsure, you should set allowExternalValidation to false, as you usually don't want external users to access your app.
+//
+// Example:
+//	func main() {
+//		idpClient := &idp.Client{}
+//		authenticate := authmiddleware.Authenticate(idpClient, tenant.SystemBaseUriFromCtx, tenant.IdFromCtx, false, logerror, loginfo)
+//		authenticateExternal := authmiddleware.Authenticate(idpClient, tenant.SystemBaseUriFromCtx, tenant.IdFromCtx, true, logerror, loginfo)
+//		mux := http.NewServeMux()
+//		mux.Handle("/hello", authenticate(helloHandler()))
+//		mux.Handle("/resource", authenticate(resourceHandler()))
+//		mux.Handle("/resource4ExternalUsers", authenticateExternal(resource4ExternalUsersHandler()))
+//	}
+//
+//	func helloHandler() http.Handler {
+//		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+//			// get user from context
+//			principal,_ := authmiddleware.PrincipalFromCtx(r.Context())
+//			// get authSessionId From context
+//			authSessionId,_ := authmiddleware.AuthSessionIdFromCtx(r.Context())
+//			fmt.Fprintf(w, "Hello %v your authsessionId is %v", principal.DisplayName, authSessionId)
+//		})
+//	}
+func Authenticate(validator Validator, getSystemBaseUriFromCtx, getTenantIdFromCtx func(ctx context.Context) (string, error), allowExternalValidation bool, logError, logInfo func(ctx context.Context, message string)) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+			ctx := req.Context()
+			authSessionId, aErr := authSessionIdFromRequest(ctx, req, logInfo)
+			if aErr != nil {
+				logError(ctx, fmt.Sprintf("error reading authSessionId from request because: %v\n", aErr))
+				http.Error(rw, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+				return
+			}
+			if authSessionId == "" {
+				acceptHeader := req.Header.Get("Accept")
+				if !isTextHtmlAccepted(acceptHeader) {
+					rw.WriteHeader(http.StatusUnauthorized)
+					rw.Header().Set("WWW-Authenticate", "Bearer")
+					return
+				}
+
+				switch req.Method {
+				case "POST", "PUT", "DELETE", "PATCH":
+					rw.WriteHeader(http.StatusUnauthorized)
+					rw.Header().Set("WWW-Authenticate", "Bearer")
+				default:
+					redirectToIdpLogin(rw, req)
+				}
+				return
+			}
+			systemBaseUri, gSBErr := getSystemBaseUriFromCtx(ctx)
+			if gSBErr != nil {
+				logError(ctx, fmt.Sprintf("error reading SystemBaseUri from context because: %v\n", gSBErr))
+				http.Error(rw, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+				return
+			}
+			tenantId, gTErr := getTenantIdFromCtx(req.Context())
```

Comment:
You got already the request context, use ctx here.

Reply:
Done